How many late days are you using for this assignment? 
0

In [1]:
# Imports and pip installations (if needed)
import pandas as pd
import numpy as np

# Part 1: Load the dataset

In [2]:
# Load the given datasets
ckd_num = pd.read_csv("chronic_kidney_disease_numerical.csv")
ckd_cat = pd.read_csv("chronic_kidney_disease_categorical.csv")

# Print the data
print("Numerical:")
print(ckd_num.head())
print("\nCategorical:")
print(ckd_cat.head())

Numerical:
   unique_id   age    bp    bgr     bu   sc    sod  pot  hemo   pcv    wbcc  \
0     203694  38.0  80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0  7300.0   
1     938027  43.0  60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0  7200.0   
2     421471  37.0  60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0  5500.0   
3     764115  70.0  90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0  8200.0   
4     240975  47.0  80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN     NaN   

   rbcc  
0   6.4  
1   5.5  
2   5.7  
3   4.5  
4   NaN  

Categorical:
   unique_id   al   su       rbc        pc         pcc          ba  htn   dm  \
0     224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes  yes   
1     992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   no   
2     308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   no   
3     450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   no   
4     881763  0.0  0.0    normal    norm

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer:
The dataset related to chronic kidney disease and uses data collected from 2 months and 400 patients. Thus, the most important column is the target one, which indicates if a patient has CKD or not. 
There are 24 patient related features, in both categorical and numerical domains. 3 Numerical features are age( years), bp (blood pressure mm/Hg), bgr(blood glucose random mgs/dl). 3 Categorical geatures are rbc (red bllood cells as 'norma' or 'abnormal'), pc (ps cell condition - 'normal' / 'abnormal'), appet (appetite quality - 'good'/'poor')

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [3]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
num_duplicate = ckd_num.duplicated().sum()
print("Total duplicate rows in numerical dataset:", num_duplicate)

# if num_duplicate > 0:
#     print("\nDuplicate rows in numerical dataset:")
#     print(ckd_num[ckd_num.duplicated(keep=False)])

ckd_num = ckd_num.drop_duplicates()
# Drop these duplicate rows

num_categorical = ckd_cat.duplicated().sum()
print("Total duplicate rows in categorical dataset:", num_categorical)

# if num_categorical > 0:
#     print("\nDuplicate rows in numerical dataset:")
#     print(ckd_cat[ckd_cat.duplicated(keep=False)])

ckd_cat = ckd_cat.drop_duplicates()

# Repeat the same for categorical dataset. Print the duplicate rows and drop them



Total duplicate rows in numerical dataset: 30
Total duplicate rows in categorical dataset: 30


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Not all unique IDs are in both datasets as the data might only incluse a subset of the categorical information. There could also be errors in data entry that resulted in not all IDs being in both. 

Answer:

In [4]:
if set(ckd_num['unique_id']) == set(ckd_cat['unique_id']):
    print("All unique IDs are in both datasets.")
else:
    print("Not all unique IDs are in both datasets.")


# Merge the two given numerical and categorical datasets based on their unique_ID.
combined_df = pd.merge(ckd_num, ckd_cat, on="unique_id", how="outer")
#Print the combined dataset
print(combined_df.head())


Not all unique IDs are in both datasets.
   unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0     102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1     103570  24.0  80.0  125.0    NaN   NaN  136.0  3.5  15.4  43.0  ...   
2     104675  65.0  80.0  215.0  133.0   2.5    NaN  NaN  13.2  41.0  ...   
3     105923  90.0  90.0  139.0   89.0   3.0  140.0  4.1  12.0  37.0  ...   
4     107297  60.0  70.0    NaN    NaN   NaN    NaN  NaN  16.4  43.0  ...   

         pc         pcc          ba  htn   dm cad appet   pe  ane  Target  
0  abnormal  notpresent  notpresent  yes   no  no  poor  yes  yes     ckd  
1    normal  notpresent  notpresent   no   no  no  good   no   no  notckd  
2    normal     present  notpresent   no  yes  no  good   no   no     ckd  
3    normal  notpresent  notpresent  yes  yes  no  good   no   no     ckd  
4    normal  notpresent  notpresent   no   no  no  good   no   no  notckd  

[5 rows x 25 columns]


## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [5]:
# Calculate the percentage of rows that contain atleast one missing value
missing_count = combined_df.isnull().any(axis=1).sum()

# Print %
missing_percentage = (missing_count / combined_df.shape[0]) * 100
print("Missing row percentage: ", missing_percentage)
# Drop these rows from the dataset
combined_df = combined_df.dropna()
# Print the Dataset
print(combined_df.head())

Missing row percentage:  61.75000000000001
   unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0     102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
6     109053  64.0  60.0  106.0   27.0   0.7  150.0  3.3  14.4  42.0  ...   
7     114717  41.0  80.0  122.0   25.0   0.8  138.0  5.0  17.1  41.0  ...   
8     118191  42.0  70.0   93.0   32.0   0.9  143.0  4.7  16.6  43.0  ...   
9     118596  74.0  60.0   88.0   50.0   0.6  147.0  3.7  17.2  53.0  ...   

         pc         pcc          ba  htn  dm cad appet   pe  ane  Target  
0  abnormal  notpresent  notpresent  yes  no  no  poor  yes  yes     ckd  
6    normal  notpresent  notpresent   no  no  no  good   no   no  notckd  
7    normal  notpresent  notpresent   no  no  no  good   no   no  notckd  
8    normal  notpresent  notpresent   no  no  no  good   no   no  notckd  
9    normal  notpresent  notpresent   no  no  no  good   no   no  notckd  

[5 rows x 25 columns]


## Part 3.4: Sort the dataset according to the Labels

In [6]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
combined_df = combined_df.sort_values(by='Target')

# Print the dataset
print(combined_df.head())

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
103     337834  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2  32.0  ...   
105     343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  24.0  ...   
108     349892  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6  37.0  ...   
110     353025  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9  31.0  ...   

           pc         pcc          ba  htn   dm  cad appet   pe  ane Target  
0    abnormal  notpresent  notpresent  yes   no   no  poor  yes  yes    ckd  
103  abnormal     present  notpresent  yes   no   no  poor  yes  yes    ckd  
105  abnormal  notpresent  notpresent  yes  yes  yes  poor  yes  yes    ckd  
108  abnormal  notpresent  notpresent  yes  yes  yes  good   no   no    ckd  
110    normal  notpresent  notpresent  yes  yes   no  poor  yes   no    ckd  

[5 rows x 25 columns]


## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [7]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']

# Print the dataset
combined_df = pd.get_dummies(combined_df, columns=categorical_columns) #get_dummies turns categorical values into numeric al by creating a new column for each possbilitu
print(combined_df.head())

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
103     337834  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2  32.0  ...   
105     343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  24.0  ...   
108     349892  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6  37.0  ...   
110     353025  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9  31.0  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no  pe_yes  ane_no  ane_yes  \
0      True    False       False        True  False    True   False     True   
103    True    False       False        True  False    True   False     True   
105   False     True       False        True  False    True   False     True   
108   False     True        True       False   True   False    True    False   
110    True    False       False        True  False    True    True    False   

     Target_ckd  Target_notckd  
0          

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: When a categorical variable with more than 2 values it can make the model believe that there is a value (ketchup) that is greater than another (mustard). To solve this, we would create dummy variables for each category so that the model does not make assumptions about the relationships.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [8]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

# Print the dataset
for col in numerical_columns:
    mean = combined_df[col].mean()
    std =  combined_df[col].std()
    lower = mean - 3 * std
    upper = mean + 3 * std
    combined_df = combined_df[(combined_df[col] >= lower) & (combined_df[col] <= upper)]

    
print(combined_df.head())


     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
105     343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
131     397388  63.0  100.0  280.0   35.0  3.2  143.0  3.5  13.0  40.0  ...   
149     438182  73.0   80.0  253.0  142.0  4.6  138.0  5.8  10.5  33.0  ...   
163     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
164     475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no  pe_yes  ane_no  ane_yes  \
105   False     True       False        True  False    True   False     True   
131   False     True        True       False   True   False    True    False   
149   False     True        True       False   True   False    True    False   
163    True    False        True       False   True   False   False     True   
164    True    False        True       False  False    True    True    False   

     Target_ckd  Target_notckd  
105        

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [9]:
from sklearn.preprocessing import StandardScaler
# Normalize the all Numerical Attributes in the dataset.
scaler = StandardScaler()
combined_df[numerical_columns] = scaler.fit_transform(combined_df[numerical_columns])

# Print the dataset
print(combined_df.head())

     unique_id       age        bp       bgr        bu        sc       sod  \
105     343710  0.810190  0.730553  1.146438  4.014305  1.690051 -0.959702   
131     397388  0.936367  2.776102  3.520558 -0.225995  1.218114  0.448743   
149     438182  1.567253  0.730553  2.921481  3.789156  2.161987 -0.431535   
163     474407 -1.713353  1.753327 -0.317973 -0.038371  0.206821 -2.720258   
164     475200  0.999456  2.776102  0.924557  0.486976  3.914895 -0.079424   

          pot      hemo       pcv  ...  cad_no  cad_yes  appet_good  \
105  1.373024 -2.913221 -2.905503  ...   False     True       False   
131 -1.303189 -0.607180 -0.590673  ...   False     True        True   
149  2.317570 -1.694935 -1.603411  ...   False     True        True   
163 -1.303189 -2.652160 -3.050180  ...    True    False        True   
164  0.428478 -2.826200 -2.616150  ...    True    False        True   

     appet_poor  pe_no  pe_yes  ane_no  ane_yes  Target_ckd  Target_notckd  
105        True  False    T

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: We should now get rid of the ID since that is not something we want to include in our modeling and predictions


In [10]:
#Remove that column
combined_df = combined_df.drop('unique_id', axis=1)

# Print the dataset
print(combined_df.head())

          age        bp       bgr        bu        sc       sod       pot  \
105  0.810190  0.730553  1.146438  4.014305  1.690051 -0.959702  1.373024   
131  0.936367  2.776102  3.520558 -0.225995  1.218114  0.448743 -1.303189   
149  1.567253  0.730553  2.921481  3.789156  2.161987 -0.431535  2.317570   
163 -1.713353  1.753327 -0.317973 -0.038371  0.206821 -2.720258 -1.303189   
164  0.999456  2.776102  0.924557  0.486976  3.914895 -0.079424  0.428478   

         hemo       pcv      wbcc  ...  cad_no  cad_yes  appet_good  \
105 -2.913221 -2.905503  0.519658  ...   False     True       False   
131 -0.607180 -0.590673  0.785483  ...   False     True        True   
149 -1.694935 -1.603411 -0.366426  ...   False     True        True   
163 -2.652160 -3.050180  1.937392  ...    True    False        True   
164 -2.826200 -2.616150 -0.233513  ...    True    False        True   

     appet_poor  pe_no  pe_yes  ane_no  ane_yes  Target_ckd  Target_notckd  
105        True  False    True   

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [11]:
# Export the dataframe to a new csv file
combined_df.to_csv("cleaned_chronic_kidney_disease.csv")
# Export the dataframe to a new json file
combined_df.to_json("cleaned_chronic_kidney_disease.json")
print(combined_df.head(15))

          age        bp       bgr        bu        sc       sod       pot  \
105  0.810190  0.730553  1.146438  4.014305  1.690051 -0.959702  1.373024   
131  0.936367  2.776102  3.520558 -0.225995  1.218114  0.448743 -1.303189   
149  1.567253  0.730553  2.921481  3.789156  2.161987 -0.431535  2.317570   
163 -1.713353  1.753327 -0.317973 -0.038371  0.206821 -2.720258 -1.303189   
164  0.999456  2.776102  0.924557  0.486976  3.914895 -0.079424  0.428478   
168  0.747101 -1.314996  3.698062 -0.188470  0.206821 -1.839980 -2.090311   
187  0.494747  1.753327  0.170164  2.475789  3.577798 -1.663925  0.743327   
395  1.314898 -0.292221  2.056147  2.063016  3.308119 -3.600537 -0.673492   
206 -0.136138 -1.314996  0.924557  1.912917  1.285533  0.096632 -0.516068   
390  0.873279 -0.292221  0.014848  0.036678  0.206821 -0.783646  0.585903   
381  0.431659  0.730553  2.056147  1.199946  1.690051 -0.607591  0.900751   
362  0.999456 -1.314996  2.610848  0.637075  1.959729 -0.607591  1.687873   

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

### Paste the markdown table here
Paste here: 
|   age    |      bp    |      bgr    |       bu      |       sc      |      sod     |      pot      |     hemo      |      pcv      |      wbcc     | cad_no | cad_yes | appet_good | appet_poor | pe_no  | pe_yes | ane_no | ane_yes | Target_ckd | Target_notckd |
|----------|------------|-------------|---------------|---------------|--------------|---------------|---------------|---------------|----------------|--------|---------|------------|------------|--------|--------|--------|---------|------------|----------------|
| 0.810190 | 0.730553   | 1.146438    | 4.014305      | 1.690051      | -0.959702    | 1.373024      | -2.913221     | -2.905503     | 0.519658       | False  | True    | False      | True       | False  | True   | False  | True    | True       | False          |
| 0.936367 | 2.776102   | 3.520558    | -0.225995     | 1.218114      | 0.448743     | -1.303189     | -0.607180     | -0.590673     | 0.785483       | False  | True    | True       | False      | True   | False  | True   | False   | True       | False          |
| 1.567253 | 0.730553   | 2.921481    | 3.789156      | 2.161987      | -0.431535    | 2.317570      | -1.694935     | -1.603411     | -0.366426      | False  | True    | True       | False      | True   | False  | True   | False   | True       | False          |
| -1.713353| 1.753327   | -0.317973   | -0.038371     | 0.206821      | -2.720258    | -1.303189     | -2.652160     | -3.050180     | 1.937392       | True   | False   | True       | False      | True   | False  | False  | True    | True       | False          |
| 0.999456 | 2.776102   | 0.924557    | 0.486976      | 3.914895      | -0.079424    | 0.428478      | -2.826200     | -2.616150     | -0.233513      | True   | False   | True       | False      | True   | False  | True   | False   | True       | False          |
| 0.747101 | -1.314996  | 3.698062    | -0.188470     | 0.206821      | -1.839980    | -2.090311     | -2.826200     | -2.760826     | 3.177909       | True   | False   | False      | True       | True   | False  | False  | True    | True       | False          |
| 0.494747 | 1.753327   | 0.170164    | 2.475789      | 3.577798      | -1.663925    | 0.743327      | -2.304078     | -2.182119     | -0.720859      | True   | False   | True       | False      | True   | False  | True   | False   | True       | False          |
| 1.314898 | -0.292221  | 2.056147    | 2.063016      | 3.308119      | -3.600537    | -0.673492     | -2.173547     | -2.326796     | 1.538654       | False  | True    | True       | False      | True   | False  | False  | True    | True       | False          |
| -0.136138| -1.314996  | 0.924557    | 1.912917      | 1.285533      | 0.096632     | -0.516068     | -1.999507     | -2.326796     | 2.912084       | True   | False   | True       | False      | True   | False  | True   | False   | True       | False          |
| 0.873279 | -0.292221  | 0.014848    | 0.036678      | 0.206821      | -0.783646    | 0.585903      | -0.781221     | -0.735350     | -0.056296      | True   | False   | True       | False      | True   | False  | True   | False   | True       | False          |
| 0.431659 | 0.730553   | 2.056147    | 1.199946      | 1.690051      | -0.607591    | 0.900751      | -1.520894     | -1.458735     | -0.277817      | True   | False   | True       | False      | True   | False  | True   | False   | True       | False          |
| 0.999456 | -1.314996  | 2.610848    | 0.637075      | 1.959729      | -0.607591    | 1.687873      | -2.130037     | -2.182119     | -0.233513      | True   | False   | False      | True       | True   | False  | True   | False   | True       | False          |
| 1.441076 | -1.314996  | -0.073905   | 3.151235      | 2.633924      | -0.783646    | 0.900751      | -1.303343     | -1.314058     | 3.177909       | True   | False   | False      | True       | True   | False  | True   | False   | True       | False          |
| 1.441076 | -0.292221  | 2.167088    | 1.537669      | 1.487792      | -1.311813    | 0.113630      | -1.738445     | -1.603411     | -1.075292      | False  | True    | True       | False      | True   | False  | True   | False   | True       | False          |
| 0.684013 | 0.730553   | 4.030883    | -0.225995     | -0.062857     | -3.248425    | -1.303189     | -1.738445     | -1.314058     | 1.272829       | True   | False   | True       | False      | True   | False  | True   | False   | True       | False          |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?
#Implement tests to verify the amount of rows/columns and stats before/after. COnvert data in batches to ensure that the LLm doesnt get overwhelmed and that you can double check the refinements. 

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [12]:
# Code to rename all the columns in the dataset
rename_mapping = {
    'age': 'Age (year)',
    'bp': 'Blood Pressure (mm/Hg)',
    'sg': 'Specific Gravity',
    'al': 'Albumin',
    'su': 'Sugar',
    'rbc': 'Red Blood Cells',
    'pc': 'Pus Cell',
    'pcc': 'Pus Cell Clumps',
    'ba': 'Bacteria',
    'bgr': 'Blood Glucose Random (mg/dl)',
    'bu': 'Blood Urea (mg/dl)',
    'sc': 'Serum Creatinine (mg/dl)',
    'sod': 'Sodium (mEq/L)',
    'pot': 'Potassium (mEq/L)',
    'hemo': 'Hemoglobin (gms)',
    'pcv': 'Packed Cell Volume',
    'wbcc': 'White Blood Cell Count (cells/cmm)',
    'rbcc': 'Red Blood Cell Count (millions/cmm)',
    'htn': 'Hypertension',
    'dm': 'Diabetes Mellitus',
    'cad': 'Coronary Artery Disease',
    'appet': 'Appetite',
    'pe': 'Pedal Edema',
    'ane': 'Anemia',
    'class': 'CKD Status'
}

# Rename the columns in the DataFrame
combined_df.rename(columns=rename_mapping, inplace=True)

# Optionally, print the updated DataFrame to verify the changes
print(combined_df.head())


     Age (year)  Blood Pressure (mm/Hg)  Blood Glucose Random (mg/dl)  \
105    0.810190                0.730553                      1.146438   
131    0.936367                2.776102                      3.520558   
149    1.567253                0.730553                      2.921481   
163   -1.713353                1.753327                     -0.317973   
164    0.999456                2.776102                      0.924557   

     Blood Urea (mg/dl)  Serum Creatinine (mg/dl)  Sodium (mEq/L)  \
105            4.014305                  1.690051       -0.959702   
131           -0.225995                  1.218114        0.448743   
149            3.789156                  2.161987       -0.431535   
163           -0.038371                  0.206821       -2.720258   
164            0.486976                  3.914895       -0.079424   

     Potassium (mEq/L)  Hemoglobin (gms)  Packed Cell Volume  \
105           1.373024         -2.913221           -2.905503   
131          -1.30

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [13]:
count_ckd_df = combined_df.groupby(['Target_ckd']).size().reset_index(name='count')
print(count_ckd_df)

   Target_ckd  count
0       False    111
1        True     22


*Describe what the above code does here*
The code counts the amount of records that relate to having CKD or not having CKD 